src/lib/transactionDecoder.ts

```diff
@@ -0,0 +1,1331 @@
+"use server";
+
+import {
+  AbiCoder,
+  FunctionFragment,
+  ParamType,
+  BytesLike,
+  ethers,
+} from "ethers";
+import axios from "axios";
+import { getAddress } from "viem";
+import {
+  mainnet,
+  sepolia,
+  optimism,
+  arbitrum,
+  base,
+  scroll,
+} from "viem/chains";
+import { unstable_cache } from "next/cache";
+import { getPublicClient } from "./viem";
+
+const DEFAULT_ETHERSCAN_API_KEY = process.env.NEXT_PUBLIC_ETHERSCAN_API_KEY;
+const SIGNATURE_DB_URL = "https://api.openchain.xyz/signature-database/v1/";
+
+const EXPLORER_DOMAINS = {
+  [mainnet.name]: "https://api.etherscan.io/api",
+  [sepolia.name]: "https://api-sepolia.etherscan.io/api",
+  [optimism.name]: "https://api-optimistic.etherscan.io/api",
+  [arbitrum.name]: "https://api.arbiscan.io/api",
+  [base.name]: "https://api.basescan.org/api",
+  [scroll.name]: "https://api.scrollscan.com/api",
+  derive: "https://explorer.derive.xyz/api",
+  cyber: "https://api.socialscan.io/cyber",
+};
+
+interface AbiItem {
+  type: string;
+  name?: string;
+  inputs?: Array<{
+    name: string;
+    type: string;
+    indexed?: boolean;
+    internalType?: string;
+    components?: any[];
+  }>;
+  outputs?: Array<{
+    name: string;
+    type: string;
+    internalType?: string;
+    components?: any[];
+  }>;
+  stateMutability?: string;
+  constant?: boolean;
+  payable?: boolean;
+  anonymous?: boolean;
+}
+
+interface DecodedParameter {
+  name: string;
+  type: string;
+  value: any;
+  components?: DecodedParameter[];
+  nestedFunction?: {
+    name: string;
+    parameters: DecodedParameter[] | Record<string, any>;
+  };
+}
+
+// -----------------------------------------------------------------------------
+// Helpers for hex encoding/decoding, offsets, and formatting
+// -----------------------------------------------------------------------------
+
+const decodeHex = (data: BytesLike): Uint8Array => {
+  if (data instanceof Uint8Array) return data;
+  if (typeof data === "string" && data.startsWith("0x")) {
+    data = data.substring(2);
+  }
+  if (typeof data === "string" && !/^([0-9a-fA-F]{2})*$/.test(data)) {
+    console.error(
+      "Invalid hex input:",
+      typeof data === "string" ? data.substring(0, 50) + "..." : String(data)
+    );
+    // Return empty array instead of throwing to prevent crashes
+    return new Uint8Array(0);
+  }
+  const result = new Uint8Array(typeof data === "string" ? data.length / 2 : 0);
+  if (typeof data === "string") {
+    for (let i = 0; i < result.length; i++) {
+      result[i] = parseInt(data.substring(i * 2, i * 2 + 2), 16);
+    }
+  }
+  return result;
+};
```

Comment:
This might be able to use ethers/viem functions to simplify the code. wouldnt replace everything but some of it

something like ethers.getBytes(data)